<a href="https://colab.research.google.com/github/vahadruya/Capstone_EDA_Global_Terrorism_Analysis/blob/main/Intermediate%20Notebooks/Day1_EDA_Global_Terrorism_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Project Name**    - Global Terrorism Dataset



##### **Project Type**    - EDA
##### **Contribution**    - Individual
##### **Name - Aditya A P**

# **Project Summary -**

Write the summary here within 500-600 words.

# **GitHub Link -**

Provide your GitHub Link here.

# **Problem Statement**


**Write Problem Statement Here.**

#### **Define Your Business Objective?**

Answer Here.

# **General Guidelines** : -  

1.   Well-structured, formatted, and commented code is required.
2.   Exception Handling, Production Grade Code & Deployment Ready Code will be a plus. Those students will be awarded some additional credits.
     
     The additional credits will have advantages over other students during Star Student selection.
       
             [ Note: - Deployment Ready Code is defined as, the whole .ipynb notebook should be executable in one go
                       without a single error logged. ]

3.   Each and every logic should have proper comments.
4. You may add as many number of charts you want. Make Sure for each and every chart the following format should be answered.
        

```
# Chart visualization code
```
            

*   Why did you pick the specific chart?
*   What is/are the insight(s) found from the chart?
* Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

5. You have to create at least 20 logical & meaningful charts having important insights.


[ Hints : - Do the Vizualization in  a structured way while following "UBM" Rule.

U - Univariate Analysis,

B - Bivariate Analysis (Numerical - Categorical, Numerical - Numerical, Categorical - Categorical)

M - Multivariate Analysis
 ]





# **Let's Begin !**

## **1. Know Your Data**

### Import Libraries

In [ ]:
# Import Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.graph_objects as go
import plotly.express as px

import warnings
warnings.filterwarnings('ignore')

### Dataset Loading

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Load Dataset
path = '/content/drive/MyDrive/ALMA BETTER/Module 4/#CAPSTONE 4 - EDA/Input data/'
df = pd.read_csv(path + 'Global Terrorism Data.csv', encoding = 'ISO-8859-1')

In [ ]:
sns.set()

### Dataset First View

In [ ]:
# Dataset First Look
df.head()

eventid  iyear  imonth  iday approxdate  extended resolution  country  \
0  197000000001   1970       7     2        NaN         0        NaN       58   
1  197000000002   1970       0     0        NaN         0        NaN      130   
2  197001000001   1970       1     0        NaN         0        NaN      160   
3  197001000002   1970       1     0        NaN         0        NaN       78   
4  197001000003   1970       1     0        NaN         0        NaN      101   

          country_txt  region  ... addnotes scite1 scite2  scite3  dbsource  \
0  Dominican Republic       2  ...      NaN    NaN    NaN     NaN      PGIS   
1              Mexico       1  ...      NaN    NaN    NaN     NaN      PGIS   
2         Philippines       5  ...      NaN    NaN    NaN     NaN      PGIS   
3              Greece       8  ...      NaN    NaN    NaN     NaN      PGIS   
4               Japan       4  ...      NaN    NaN    NaN     NaN      PGIS   

   INT_LOG  INT_IDEO INT_MISC INT_ANY  related  
0        0         0        0       0      NaN  
1        0         1        1       1      NaN  
2       -9        -9        1       1      NaN  
3       -9        -9        1       1      NaN  
4       -9        -9        1       1      NaN  

[5 rows x 135 columns]

### Dataset Rows & Columns count

In [ ]:
# Dataset Rows & Columns count
print(f'Number of rows in the dataset: {df.shape[0]}')
print(f'Number of columns  in the dataset: {df.shape[1]}')

Number of rows in the dataset: 181691
Number of columns  in the dataset: 135


### Dataset Information

In [ ]:
# Dataset Info
df.info(verbose = True, show_counts = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 181691 entries, 0 to 181690
Data columns (total 135 columns):
 #    Column              Non-Null Count   Dtype  
---   ------              --------------   -----  
 0    eventid             181691 non-null  int64  
 1    iyear               181691 non-null  int64  
 2    imonth              181691 non-null  int64  
 3    iday                181691 non-null  int64  
 4    approxdate          9239 non-null    object 
 5    extended            181691 non-null  int64  
 6    resolution          2220 non-null    object 
 7    country             181691 non-null  int64  
 8    country_txt         181691 non-null  object 
 9    region              181691 non-null  int64  
 10   region_txt          181691 non-null  object 
 11   provstate           181270 non-null  object 
 12   city                181257 non-null  object 
 13   latitude            177135 non-null  float64
 14   longitude           177134 non-null  float64
 15   specificity    

#### Duplicate Values

In [ ]:
# Dataset Duplicate Value Count
print(f'Number of duplicated rows in the dataset: {df.duplicated().sum()}')

Number of duplicated rows in the dataset: 0


It is expected that **eventid** column should have all unique values, as it represents a unique event of terrorism.

In [ ]:
df['eventid'].duplicated().sum()

0

#### Missing Values/Null Values

In [ ]:
# Missing Values/Null Values Count
print(f'There are {df.isna().sum().sum()} missing values in the dataset\n\nThe columns which have missing values are:')
dfna = df.isna().sum()
print(dfna[dfna != 0])

There are 13853997 missing values in the dataset

The columns which have missing values are:
approxdate    172452
resolution    179471
provstate        421
city             434
latitude        4556
               ...  
addnotes      153402
scite1         66191
scite2        104758
scite3        138175
related       156653
Length: 106, dtype: int64


In [ ]:
# Visualizing the missing values of only columns with missing values
# plt.figure(figsize = (13, 10))
# sns.heatmap(df[dfna[dfna != 0].index].isna())

### What did you know about your dataset?

On a first look at the dataset, it is found that
*   There are **181691 rows and 135 columns**, out of which one is **eventid**, which is a primary key to the dataset.
*   There are **no duplicated data** in the dataset, and all values in **eventid** column are **unique**
*   There are **several missing values** in the dataset, since many of the columns are markers or counters which display null values when not necessary.
*   Due to the extremely large width of the dataset, only necessary columns will be chosen based on intelligent selection for Data Analysis. Null values shall be dealt with on this select features only.

## **2. Understanding Your Variables**

Since there are several columns in the dataset, it is cumbersome to analyse each of them in this section. Instead, part of **Data Wrangling** shall be done here so that only certain variables (deemed important for a proper EDA of the dataset) shall be selected.

First, those columns are filtered out which contain a large number of missing values (~15-20% of the column)

In [ ]:
# Computing the percentage of missing values in each column
dfnan = df.isna().sum()*100/df.shape[0]

# Printing columns with less than 15% missing values
dfnan[dfnan < 15].index

Index(['eventid', 'iyear', 'imonth', 'iday', 'extended', 'country',
       'country_txt', 'region', 'region_txt', 'provstate', 'city', 'latitude',
       'longitude', 'specificity', 'vicinity', 'crit1', 'crit2', 'crit3',
       'doubtterr', 'multiple', 'success', 'suicide', 'attacktype1',
       'attacktype1_txt', 'targtype1', 'targtype1_txt', 'targsubtype1',
       'targsubtype1_txt', 'target1', 'natlty1', 'natlty1_txt', 'gname',
       'guncertain1', 'individual', 'weaptype1', 'weaptype1_txt',
       'weapsubtype1', 'weapsubtype1_txt', 'nkill', 'nwound', 'property',
       'ishostkid', 'dbsource', 'INT_LOG', 'INT_IDEO', 'INT_MISC', 'INT_ANY'],
      dtype='object')

In [ ]:
# Selecting necessary columns for EDA
cols = [
        'eventid', 'iyear', 'imonth', 'iday', 'country_txt', 'region_txt', 'provstate',
        'city', 'latitude', 'longitude', 'success', 'suicide', 'attacktype1_txt',
        'targtype1_txt', 'gname', 'weaptype1_txt', 'nkill', 'nwound', 'summary', 'motive'
       ]

data = df[cols].copy()

# Renaming columns appropriately
col_rename = {
              'iyear':'year', 'imonth':'month', 'iday':'day', 'country_txt':'country', 'region_txt':'region', 'gname':'group',
              'provstate':'state', 'attacktype1_txt':'attack_type', 'targtype1_txt':'target', 'weaptype1_txt':'weapon_type',
              'nkill':'killed', 'nwound':'wounded'
             }

data.rename(columns = col_rename, inplace = True)

*   Out of the above selected columns, the following are used only for referencing or if and when studying particular/specific cases, and are not primarily involved in the EDA:
  *   **month**
  *   **day**
  *   **provstate**
  *   **city**
  *   **latitude**
  *   **longitude**
  *   **summary**
  *   **motive**


In [ ]:
# New missing values Count
print(f'There are {data.isna().sum().sum()} missing values in the new dataset\n\nThe columns which have missing values are:')
dfna = data.isna().sum()
print(dfna[dfna != 0])

There are 233851 missing values in the new dataset

The columns which have missing values are:
state           421
city            434
latitude       4556
longitude      4557
killed        10313
wounded       16311
summary       66129
motive       131130
dtype: int64


In [ ]:
# Dataset Columns
print(f'There are {len(data.columns)} columns in the dataset.\nThey are:')
print(*data.columns, sep = '\n')

There are 20 columns in the dataset.
They are:
eventid
year
month
day
country
region
state
city
latitude
longitude
success
suicide
attack_type
target
group
weapon_type
killed
wounded
summary
motive


In [ ]:
# Dataset Describe
data.describe()

eventid           year          month            day  \
count  1.816910e+05  181691.000000  181691.000000  181691.000000   
mean   2.002705e+11    2002.638997       6.467277      15.505644   
std    1.325957e+09      13.259430       3.388303       8.814045   
min    1.970000e+11    1970.000000       0.000000       0.000000   
25%    1.991021e+11    1991.000000       4.000000       8.000000   
50%    2.009022e+11    2009.000000       6.000000      15.000000   
75%    2.014081e+11    2014.000000       9.000000      23.000000   
max    2.017123e+11    2017.000000      12.000000      31.000000   

            latitude     longitude        success        suicide  \
count  177135.000000  1.771340e+05  181691.000000  181691.000000   
mean       23.498343 -4.586957e+02       0.889598       0.036507   
std        18.569242  2.047790e+05       0.313391       0.187549   
min       -53.154613 -8.618590e+07       0.000000       0.000000   
25%        11.510046  4.545640e+00       1.000000       0.000000   
50%        31.467463  4.324651e+01       1.000000       0.000000   
75%        34.685087  6.871033e+01       1.000000       0.000000   
max        74.633553  1.793667e+02       1.000000       1.000000   

              killed        wounded  
count  171378.000000  165380.000000  
mean        2.403272       3.167668  
std        11.545741      35.949392  
min         0.000000       0.000000  
25%         0.000000       0.000000  
50%         0.000000       0.000000  
75%         2.000000       2.000000  
max      1570.000000    8191.000000

From a brief description of the dataset, it can be observed that this dataset contains information of terrorist events from the year **1970** till **2017**. Also, as many as **1570** people have been killed in one particular terrorist event.

### Variables Description

The **[Global Terrorism Dataset](https://www.start.umd.edu/gtd/)** (or GTD for short) contains information on approximately 2,00,000 terrorist events across the globe spanning several years. Since the dataset is vast in terms of features for each terrorist event, only select features are considered here for EDA, and only their descriptions are provided below. Detailed description of each feature can be found in this [document](https://drive.google.com/file/d/1TLRkQDyLStykrjifbCqoMNVlXuk3LBAn/view?usp=sharing):

**NOTE**: The variables colour-coded with green background below are the "**primary variables**", i.e., variables which are primarily being used for the EDA (FILL)

### Check Unique Values for each variable.

In [ ]:
# Check Unique values for primary categorical variables
vars = ['year', 'country', 'success', 'suicide', 'attack_type', 'target', 'group', 'weapon_type']
for var in vars:
  to_print = 5
  suff = f'etc.. {data[var].nunique()} values' if data[var].nunique() > to_print else ''
  print(f'Unique values in "{var}" are: {sorted(data[var].unique())[:to_print]} {suff}')

Unique values in "year" are: [1970, 1971, 1972, 1973, 1974] etc.. 47 values
Unique values in "country" are: ['Afghanistan', 'Albania', 'Algeria', 'Andorra', 'Angola'] etc.. 205 values
Unique values in "success" are: [0, 1] 
Unique values in "suicide" are: [0, 1] 
Unique values in "attack_type" are: ['Armed Assault', 'Assassination', 'Bombing/Explosion', 'Facility/Infrastructure Attack', 'Hijacking'] etc.. 9 values
Unique values in "target" are: ['Abortion Related', 'Airports & Aircraft', 'Business', 'Educational Institution', 'Food or Water Supply'] etc.. 22 values
Unique values in "group" are: ['1 May', '14 K Triad', '14 March Coalition', '14th of December Command', '15th of September Liberation Legion'] etc.. 3537 values
Unique values in "weapon_type" are: ['Biological', 'Chemical', 'Explosives', 'Fake Weapons', 'Firearms'] etc.. 12 values


It can be roughly observed that the dataset contains information of terrorist attacks
* from **1970 to 2017**
* of **3537** distinct **organisations**
* using **various methods of attack** (such as **Bombing** and **Hijacking**)
* and a variety of **weapons** (such as **Explosives** and **Firearms**).
* The organisations also had quite a range of **targets** from **Airports** to **Educational Institutions**, some of them involving suicide attacks.

## **3. Data Wrangling**

Part of **Data Wrangling** has already been performed in the previous section in order to trim down the large number of variables. In this section, (FILL)

### 1. Handling missing values

*   As stated earlier, the dataset is segregated into two - **primary** and **secondary** features. Out of the total 21 columns, 8 are secondary and has been selected as a reference which may be useful only for a secondary EDA, and not primarily. Hence, none of the missing values from the secondary columns have to be imputed, which preserves data.
*  Hence, out of the 8 columns which contain missing values, only two are left which are primary variables - viz., **killed** and **wounded**.

In [ ]:
# Checking missing values in nkill and nwound
print('Missing values in killed and wounded:')
print(data[['killed', 'wounded']].isna().sum())

tot_rows = data.shape[0]
n_rows = data.dropna(subset = ['killed', 'wounded']).shape[0]
print('\nNumber of rows of data lost on dropping these null values:', end = ' ')
print(f'{tot_rows - n_rows} or {round((tot_rows - n_rows)*100/tot_rows, 2)}%')

Missing values in killed and wounded:
killed     10313
wounded    16311
dtype: int64

Number of rows of data lost on dropping these null values: 16874 or 9.29%


In [ ]:
# Plotting year-wise terrorism events with and without NA of nkill and nwound

def line_plots(to_plot_dfs, names = [None], figure_ = True, markers = True):
  '''Returns a figure which has the line-plot of plotly. Inputs are:
  the dataframes to plot, stored in a list; the names of each of the plots, also as a list;
  and a boolean to indicate whether the figure or the data to plot is to be returned'''
  to_plot = []
  if markers == True:
    for i in range(len(to_plot_dfs)):
      to_plot.append(go.Scatter(x = to_plot_dfs[i].index, y = to_plot_dfs[i].values, mode = 'lines+markers', marker = dict(symbol = 'circle', size = 8), name = names[i]))
  else:
    for i in range(len(to_plot_dfs)):
      to_plot.append(go.Scatter(x = to_plot_dfs[i].index, y = to_plot_dfs[i].values, name = names[i], mode = 'lines'))
  if figure_ == True:
    fig = go.Figure(data = to_plot)
    return fig
  return to_plot

yearly_df = data['year'].value_counts().sort_index()
yearly_dropna_df = data.dropna(subset = ['killed', 'wounded'])['year'].value_counts().sort_index()
names = ['With NA of killed and wounded', 'Dropping NA of killed and wounded']

fig = line_plots([yearly_df, yearly_dropna_df], names, True)
fig.update_layout(title = 'Yearly terrorism events - with and without dropping NA of killed and wounded',
                  yaxis_title = 'Number of terrorist events',
                  xaxis_title = 'Year',
                  autosize = False, width = 1300, height = 600)
fig.show()

*   It can be seen that there is signifcant loss of data on dropping the null values of **killed** and **wounded** (~9.3%) which is also reflected in a drop of terrorist activities across each year.
*   Since each row in the dataset corresponds to a terrorist event, any loss of information will have crucial effects in the analysis. Also, imputation of these missing values with any other value does not seem logical since (as per the document linked) "***this field remains blank when the figure is not reported or it is too vague to be of use***"
*   Hence, these null values shall not be dropped or tampered with, and the entire dataset will be used whenever **killed** and **wounded** are not included for analysis, while these null values will be ignored when the same 2 variables will be used for EDA.

### Data Wrangling Code

In [ ]:
# Creating a column for casualties
data['casualties'] = data['killed'] + data['wounded']
data['weapon_type'] = data['weapon_type'].replace('Vehicle (not to include vehicle-borne explosives, i.e., car or truck bombs)', 'Vehicle')

In [ ]:
# Creating a dataframe for number of events per year
yearly_df = data['year'].value_counts().sort_index()

# Creating a dataframe grouped by year and aggregating other variables by addition
yearly_sum = data.groupby('year').sum()

In [ ]:
# Creating a dataframe for number of events per region
region_df = data['region'].value_counts()
# Creating a dataframe grouped by region and aggregating other variables by addition
region_sum = data.groupby('region').sum()

# Repeating above steps of region column for country, group, attack type, weapon type and targets
country_df = data['country'].value_counts()
country_sum = data.groupby('country').sum()

group_df = data['group'].value_counts()
group_sum = data.groupby('group').sum()

attack_df = data['attack_type'].value_counts()
attack_sum = data.groupby('attack_type').sum()

weapon_df = data['weapon_type'].value_counts()
weapon_sum = data.groupby('weapon_type').sum()

target_df = data['target'].value_counts()
target_sum = data.groupby('target').sum()

In [ ]:
# Creating dataframe with number of events for each weapon type for top 3 attack types
attack_weapon_df = data.groupby(['attack_type', 'weapon_type']).count()['eventid']
top3attack_weapon = attack_weapon_df.loc[attack_df[:3].index].groupby('attack_type', group_keys = False).apply(lambda x: x.sort_values(ascending = False))
top3attack_weapon = top3attack_weapon.to_frame().rename(columns = {'eventid': 'number of events'})

# Creating dataframe with number of events for each attack type for top 5 targets
target_attack_df = data.groupby(['target', 'attack_type']).count()['eventid']
top5target_attack = target_attack_df.loc[target_df[:5].index].groupby('target', group_keys = False).apply(lambda x: x.sort_values(ascending = False))
top5target_attack = top5target_attack.to_frame().rename(columns = {'eventid': 'number of events'})

### What all manipulations have you done and insights you found?

Answer Here.

## **4. Data Vizualization, Storytelling & Experimenting with charts : Understand the relationships between variables**

### 1. Global Analysis

#### Chart 1 - Yearly analysis

In [ ]:
# Plotting the distribution of terrorist events by year
fig = line_plots([yearly_df])
fig.update_layout(title = 'Yearly terrorism events', yaxis_title = 'Number of terrorist events', xaxis_title = 'Year',
                  autosize = False, width = 1100, height = 600)
fig.show()

In [ ]:
# Dataframe for number of casualties by year
cas_cols = ['killed', 'wounded', 'casualties']
yearly_sum[cas_cols].sort_values(cas_cols, ascending = False).head()

killed  wounded  casualties
year                              
2014  44490.0  41128.0     68853.0
2015  38853.0  44043.0     72833.0
2016  34871.0  40001.0     69225.0
2017  26445.0  24927.0     48208.0
2013  22273.0  37688.0     57679.0

In [ ]:
# Plotting the number of killed and wounded by year
yr_kw_list = [yearly_sum[var] for var in cas_cols[:-1]]
fig = line_plots(yr_kw_list, ['Number of people killed', 'Number of people wounded'])
fig.update_layout(title = 'Yearly casualties due to terrorism', yaxis_title = 'Number of people', xaxis_title = 'Year',
                  autosize = False, width = 1300, height = 600)
fig.show()

In [ ]:
# Plot of number of people killed, wounded, and total casualties
for var in cas_cols:
  text = f'people {var}' if var != 'casualties' else var
  # Plots
  line_traces = line_plots([yearly_sum[var]], [f'Number of {text}'], figure_ = False)
  avg_trace = go.Bar(x = yearly_sum.index, y = (yearly_sum[var] / yearly_df), yaxis = 'y2', name = f'Average {text} per event', marker = dict(opacity = 0.5, color = '#00CC96'))
  fig = go.Figure(data = [*line_traces, avg_trace])
  # Updating the layout
  fig.update_layout(title = f'Yearly total and average number of {text}', xaxis_title = 'Year',
                    yaxis_title = 'Number of people', yaxis2 = dict(title = 'Avg number of people', overlaying = 'y', side = 'right'),
                    autosize = False, width = 1400, height = 600)
  fig.show()

In [ ]:
# Dataframe for number of succesful and suicide events by year
cols = ['success', 'suicide']
yearly_sum[cols].sort_values(cols, ascending = False).head()

success  suicide
year                  
2014    15015      744
2015    12676      922
2016    10975      985
2013    10484      622
2017     8652      844

In [ ]:
# Plotting on successful and suicide events
texts = ['successful events', 'suicide events']
vars = ['success', 'suicide']
for text, var in zip(texts, vars):
  to_plot_df = [yearly_sum[var], yearly_df, ] if var == 'success' else [yearly_sum[var]]
  yrange = [78, 100] if var == 'success' else [0, 12]
  line_traces = line_plots(to_plot_df, [f'Number of {text}', 'Number of events'], figure_ = False)
  success_trace = go.Bar(x = yearly_sum.index, y = (yearly_sum[var]/yearly_df * 100), yaxis = 'y2', name=f'Percentage of {text}', marker=dict(opacity = 0.4, color = '#00CC96'))
  fig = go.Figure(data = [*line_traces, success_trace])
  fig.update_layout(title = f'Yearly {text}', xaxis_title = 'Year', yaxis_title = 'Number of events', yaxis2 = dict(title = f'Percentage', overlaying = 'y', side = 'right'),
                    yaxis2_range = yrange, autosize = False, width = 1400, height = 600)
  fig.show()

##### 1. Why did you pick the specific chart?

Library used for plots:
*   For EDA in this project, **plotly** has been used primarily instead of **matplotlib** (despite the simplicity of the latter) because of the interactive interface of plotly. Interactive visualizations allow users to explore and interact with data directly in the plot. This can be especially helpful when dealing with large datasets, as users can zoom, pan, hover over data points, and more to extract insights.

Type of plots:
*   For the temporal visualization, **line plots** have been used for their simplicity and ability in showing trends and changes over time. The continuous lines connect data points in chronological order, helping to emphasize the temporal progression.
*   For displaying percentages or averages overlaying with the temporal lineplots, the **box-plots** with adjusted opacity have been used to avoid cluttering and for improving interpretability.



##### 2. What is/are the insight(s) found from the chart?

This chart is a basic general temporal chart to visualize the dataset from a higher level. The features **killed, wounded, casualties, success** and **suicide** are parameters to measure the impact of a terrorist attack, and hence shall be used in every Chart.
1.   The **first plot** shows the **number of terrorist events across the years**. It can be observed that the number of terrorist events **gradually increased** till **1992**, after which it **dropped** till **2004**. Then there was a **steep rise** in number of terrorist activities till it **peaked in 2014**, after which it has **considerably decreased** till **2017**.
2.   Keeping in line with this trend, the **second plot** shows a **similar pattern** for number of people **killed and wounded**. But there is a **local peak at 2001**, the year which had less number of terrorist events. This means high number of casualties in few number of attacks, possibly due to the famous 9/11 in New York, This will be analysed in the next Chart.
3.   Following this, the **third plot** shows the **average number of killed/wounded/casualties per event every year** as an overlay with the total number every year. It can be seen that the **average number of casualties per event is the maximum in 2001** followed by 2004. The former is possibly the effect of the 9/11 attacks. In 2004 the number of attacks were less, but casualties per attack were high. **The number of casualties per attack has significantly decreased since then, despite increase in number of events.**
4.   The **fourth plot** is similar to the third one, but for events which are deemed as **success** or events which involve **suicide attacks**. The number of successful attacks has also reduced significantly in 21st century, number of suicide attacks has increased manifold which was almost rare in 20th century.


**NOTE**: **success** is defined (as in the document in variable descriptions in previous section) based on the tangible effects of the attack, and not in terms of the larger goals/motives of the perpetrators.

#### Chart 2 - Demographic (Regional) analysis

In [ ]:
# Plotting region-wise distribution of terrorist attacks
texts = ['terrorist events', 'suicide terrorist events', 'casualties']
for var, text in zip([None, 'suicide', 'casualties'], texts):
  region_plot_df = region_df if var == None else region_sum[var].sort_values(ascending = False)
  region_plot = go.Bar(x = region_plot_df.index, y = region_plot_df.values)
  fig = go.Figure(data = region_plot)
  fig.update_layout(title = f'Region-wise distribution of {text}', xaxis_title = 'Region', yaxis_title = f'Number of {text.split(" ")[-1]}',
                    autosize = False, width = 1200, height = 600)
  fig.show()

In [ ]:
def multi_cat_lineplot(names_toplot, toplot_dict, title, var = ''):
  events = True if var == '' else False
  if events == True:
    categ_toplot = list({k: toplot_dict[k] for k in names_toplot}.values())
  else:
    categ_toplot = list({k: toplot_dict[k][var] for k in names_toplot}.values())

  fig = line_plots(categ_toplot, names_toplot, markers = False)
  var = 'terrorism' if var == '' else var
  var = var + ' events' if var != 'casualties' else var
  fig.update_layout(title = f'Yearly {var} across different {title}', yaxis_title = f'Number of {var}', xaxis_title = 'Year',
                    autosize = False, width = 1400, height = 650)
  fig.show()

# Plotting region-wise distribution of terrorist attacks across each year
region_names = data['region'].unique()[::-1]
region_names_tp = [region for region in region_names if region not in ['East Asia', 'Central Asia', 'Australasia & Oceania']]
region_year_dict = {region: data[data['region'] == region]['year'].value_counts().sort_index() for region in region_names}

multi_cat_lineplot(region_names_tp, region_year_dict, 'regions')

In [ ]:
# Plotting region-wise distribution of casualties across each year
region_yr_sum_dict = {region: data[data['region'] == region].groupby('year').sum() for region in region_names}
multi_cat_lineplot(region_names_tp, region_yr_sum_dict, 'regions', 'casualties')

In [ ]:
# 9/11 attacks
data[(data['region'] == 'North America') & (data['year'] == 2001) & (data['month'] == 9) & (data['day'] == 11)][cas_cols + ['country', 'city']]

killed  wounded  casualties        country           city
73126  1384.0   8190.0      9574.0  United States  New York City
73127  1383.0   8191.0      9574.0  United States  New York City
73128   190.0    106.0       296.0  United States      Arlington
73129    44.0      6.0        50.0  United States    Shanksville

In [ ]:
# Plotting region-wise distribution of casualties across each year
multi_cat_lineplot(region_names_tp, region_yr_sum_dict, 'regions', 'suicide')

##### 1. Why did you pick the specific chart?

*   To visualise the distribution of the total number events, casualties and suicide attacks across regions (cumulated over all time), the **bar plots** have been used. When temporal factor comes into play, the lineplots have been bought back.
*   This is because unlike the line plots, barplots do not intend to highlight the trend between different values of the x-coordinate, which is required in temporal analysis. Instead, the barplot helps the user analyse the categories independently.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart 3 - Demographic (National) analysis

In [ ]:
# Top 20 countries in terms of terrorist events, suicide events and casualties
texts = ['terrorist events', 'casualties', 'suicide terrorist events']
for var, text in zip([None, 'casualties', 'suicide'], texts):
  country_plot_df = country_df.head(20) if var == None else country_sum[var].sort_values(ascending = False).head(20)
  country_plot = go.Bar(x = country_plot_df.index, y = country_plot_df.values)
  fig = go.Figure(data = country_plot)
  fig.update_layout(title = f'Country-wise distribution of {text} (top-20)', xaxis_title = 'Country', yaxis_title = f'Number of {text.split(" ")[-1]}',
                    autosize = False, width = 1200, height = 600)
  fig.show()

In [ ]:
# Choropleth map coloured by number of terrorist events
def plot_world(df, title_):
  fig = px.choropleth(df, locations = df.index, locationmode = 'country names', color = df.values,
                    hover_name = df.index, projection = 'natural earth', color_continuous_scale="Viridis")
  fig.update_layout(title = title_)
  fig.show()

plot_world(country_df, 'Map of number of terrorist events')

In [ ]:
# Choropleth map coloured by number of casualties
country_cas = country_sum['casualties'].sort_values(ascending = False)
plot_world(country_cas, 'Map of number of casualties')

In [ ]:
# Choropleth map coloured by number of suicide events
country_sui = country_sum['suicide'].sort_values(ascending = False)
plot_world(country_sui, 'Map of number of suicide events')

In [ ]:
# Yearly terrorism events across top 10 countries
country_names = list(country_df.head(10).index)
country_year_dict = {country: data[data['country'] == country]['year'].value_counts().sort_index() for country in country_names}

multi_cat_lineplot(country_names, country_year_dict, 'countries')

In [ ]:
# Yearly terrorism casualties across top 10 countries
country_names = list(country_sum['casualties'].sort_values(ascending = False).head(10).index)
country_yr_sum_dict = {country: data[data['country'] == country].groupby('year').sum() for country in country_names}

multi_cat_lineplot(country_names, country_yr_sum_dict, 'countries', 'casualties')

In [ ]:
# Yearly suicide events across top 10 countries
country_names = list(country_sum['suicide'].sort_values(ascending = False).head(10).index)
country_yr_sum_dict = {country: data[data['country'] == country].groupby('year').sum() for country in country_names}

multi_cat_lineplot(country_names, country_yr_sum_dict, 'countries', 'suicide')

##### 1. Why did you pick the specific chart?

*   Similar to the previous chart, the **line and barplots** have been used to visualize the number of events, casualties and suicide attacks across each country. Lineplots have been used when there is temporal factor.
*   Since there is large number of countries, all of them can not be displayed in the bar or line plots. Hence, the **choropleth** plots have been used to visualize in a global map.


##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart 4 - Analysis of Terrorist Organisations

In [ ]:
group_df.head()

Unknown                                             82782
Taliban                                              7478
Islamic State of Iraq and the Levant (ISIL)          5613
Shining Path (SL)                                    4555
Farabundo Marti National Liberation Front (FMLN)     3351
Name: group, dtype: int64

A vast majority of the events are by **Unknown** groups, hence for the sake of visualization they shall be put away.

In [ ]:
texts = ['terrorist events', 'suicide terrorist events', 'casualties']
for var, text in zip([None, 'suicide', 'casualties'], texts):
  group_plot_df = group_df[1:].head(20) if var == None else group_sum[var].sort_values(ascending = False)[1:].head(20)
  group_plot = go.Bar(x = group_plot_df.index, y = group_plot_df.values)
  fig = go.Figure(data = group_plot)
  fig.update_layout(title = f'Terrorist group-wise distribution of {text}', xaxis_title = 'Terrorist group', yaxis_title = f'Number of {text.split(" ")[-1]}',
                    autosize = False, width = 1200, height = 600)
  fig.show()

In [ ]:
var = 'success'
groups_toplot = group_sum[var].sort_values(ascending = False)[1:].head(20)

group_success_pct = groups_toplot / group_df[groups_toplot.index] * 100
group_success_pct.sort_values(ascending = False).to_frame().rename(columns = {0: 'Percentage of success'}).head(10)

Percentage of success
group                                                                  
Nicaraguan Democratic Force (FDN)                             98.994413
Farabundo Marti National Liberation Front (FMLN)              98.985377
Liberation Tigers of Tamil Eelam (LTTE)                       95.267746
Shining Path (SL)                                             95.214050
Communist Party of India - Maoist (CPI-Maoist)                94.462194
National Liberation Army of Colombia (ELN)                    94.170404
Kurdistan Workers' Party (PKK)                                94.155844
Tehrik-i-Taliban Pakistan (TTP)                               93.782383
Boko Haram                                                    92.928040
Revolutionary Armed Forces of Colombia (FARC)                 92.561319

In [ ]:
# Plotting success events and percentage of success for top 10 groups
def pct_success_plot(df_toplot, pct_df, var, pct_range):
  bar_trace = go.Bar(x = df_toplot.index, y = (df_toplot), yaxis = 'y2', name = 'Total successful events', marker = dict(opacity = 0.5, color = '#00CC96'))
  line_trace = line_plots([pct_df], names = ['Percentage of success'], figure_ = False, markers = False)

  fig = go.Figure(data = [*line_trace, bar_trace])
  fig.update_layout(title = f'Percentage and Total successful events for each {var}', xaxis_title = var,
                    yaxis_title = 'Percentage of success', yaxis2 = dict(title = 'Total success events', overlaying = 'y', side = 'right'),
                    yaxis_range = pct_range, autosize = False, width = 1400, height = 650)
  fig.show()

pct_success_plot(groups_toplot, group_success_pct, 'group', [70, 100])

In [ ]:
# Yearly terrorism events across top 10 groups
group_names = list(group_df[1:].head(10).index)
group_year_dict = {group: data[data['group'] == group]['year'].value_counts().sort_index() for group in group_names}
multi_cat_lineplot(group_names, group_year_dict, 'groups')

In [ ]:
# Yearly terrorism casualties across top 10 groups
for var in ['casualties', 'success', 'suicide']:
  group_names = list(group_sum[var].sort_values(ascending = False)[1:].head(10).index)
  group_yr_sum_dict = {group: data[data['group'] == group].groupby('year').sum() for group in group_names}

  multi_cat_lineplot(group_names, group_yr_sum_dict, 'groups', var)

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart 5 - Analysis of Attack Types and Weapon Types

In [ ]:
for df, text in zip([attack_df, weapon_df], ['Attack Type', 'Weapon Type']):
  fig = go.Figure(data = go.Pie(values = df.values, labels = df.index))
  fig.update_layout(title = f'Proportions of each {text} on total terrorist events', autosize = False, width = 1300, height = 550)
  fig.show()

In [ ]:
for df, text in zip([attack_sum, weapon_sum], ['Attack Type', 'Weapon Type']):
  sui_df = df['suicide'].sort_values(ascending = False)
  fig = go.Figure(data = go.Pie(values = sui_df.values, labels = sui_df.index))
  fig.update_layout(title = f'Proportions of each {text} for suicide attacks', autosize = False, width = 1300, height = 550)
  fig.show()

In [ ]:
# Plotting success events and percentage of success for attack types
attack_success = attack_sum['success'].sort_values(ascending = False)
attack_success_pct = attack_success / attack_df[attack_success.index] * 100
pct_success_plot(attack_success, attack_success_pct, 'attack type', [30, 100])

In [ ]:
for df, text in zip([attack_sum, weapon_sum], ['Attack Type', 'Weapon Type']):
  cas_df = df['casualties'].sort_values(ascending = False)
  fig = go.Figure(data = go.Bar(x = cas_df.index, y = cas_df.values))
  fig.update_layout(title = f'Casualties per {text}', xaxis_title = text, yaxis_title = 'Number of casualties}',
                    autosize = False, width = 1200, height = 600)
  fig.show()

In [ ]:
top3_attacks = {index[0] for index in top3attack_weapon.index}
for var in top3_attacks:
  fig = go.Figure(data = go.Pie(values = top3attack_weapon.loc[var]['number of events'], labels = top3attack_weapon.loc[var].index))
  fig.update_layout(title = f'Proportions of each weapon type for {var} attacks', autosize = False, width = 1300, height = 550)
  fig.show()

In [ ]:
# Yearly terrorism events across
attack_yr_dict = {attack_type: data[data['attack_type'] == attack_type]['year'].value_counts().sort_index() for attack_type in list(attack_df.index)}
multi_cat_lineplot(list(attack_df.index), attack_yr_dict, 'Attack Types')

In [ ]:
# Yearly terrorism casualties across different attack types
for var in ['casualties', 'success']:
  attack_yr_sum_dict = {attack_type: data[data['attack_type'] == attack_type].groupby('year').sum() for attack_type in list(attack_df.index)}
  multi_cat_lineplot(list(attack_df.index), attack_yr_sum_dict, 'attack types', 'casualties')

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart 6 - Analysis of Target Types

In [ ]:
df = target_df
fig = go.Figure(data = go.Pie(values = df.values, labels = df.index))
fig.update_layout(title = f'Proportions of each target of terrorists', autosize = False, width = 1300, height = 550)
fig.show()

In [ ]:
sui_df = target_sum['suicide'].sort_values(ascending = False)
fig = go.Figure(data = go.Pie(values = sui_df.values, labels = sui_df.index))
fig.update_layout(title = f'Proportions of targets of suicide attacks', autosize = False, width = 1300, height = 550)
fig.show()

In [ ]:
# Plotting success events and percentage of success for attack types
target_success = target_sum['success'].sort_values(ascending = False)
target_success_pct = target_success / target_df[target_success.index] * 100
pct_success_plot(target_success, target_success_pct, 'target', [30, 100])

In [ ]:
target_cas_df = target_sum['casualties'].sort_values(ascending = False)
fig = go.Figure(data = go.Bar(x = target_cas_df.index, y = target_cas_df.values))
fig.update_layout(title = 'Casualties per target', xaxis_title = text, yaxis_title = 'Number of casualties}',
                  autosize = False, width = 1200, height = 600)
fig.show()

In [ ]:
# Yearly terrorism casualties across top targets
target_names = list(target_df.head(10).index)
target_yr_dict = {target: data[data['target'] == target]['year'].value_counts().sort_index() for target in target_names}
multi_cat_lineplot(target_names, target_yr_dict, 'targets')

In [ ]:
# Yearly terrorism casualties across top targets
for var in ['casualties', 'success', 'suicide']:
  target_names = list(target_sum[var].sort_values(ascending = False).head(10).index)
  target_yr_sum_dict = {target: data[data['target'] == target].groupby('year').sum() for target in target_names}
  multi_cat_lineplot(target_names, target_yr_sum_dict, 'targets', var)

In [ ]:
top5_targets = {index[0] for index in top5target_attack.index}
for var in top5_targets:
  fig = go.Figure(data = go.Pie(values = top5target_attack.loc[var]['number of events'], labels = top5target_attack.loc[var].index))
  fig.update_layout(title = f'Distribution of types of attack for {var} target', autosize = False, width = 1300, height = 550)
  fig.show()

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

### 2. Local Analysis

In [ ]:
countries = country_df.head(4).index

In [ ]:
country_dict = {country: data[data['country'] == country] for country in countries}
c_yearly_sum = {country: country_dict[country].groupby('year').sum() for country in countries}

c_attack_dict = {country: country_dict[country]['attack_type'].value_counts() for country in countries}
c_weapon_dict = {country: country_dict[country]['weapon_type'].value_counts() for country in countries}
c_target_dict = {country: country_dict[country]['target'].value_counts() for country in countries}
c_group_dict = {country: country_dict[country]['group'].value_counts() for country in countries}

#### Chart 8 - Iraq

Events and success by year

kill/wound by year

groups by year

map



In [ ]:
def generate_country_plots(cname):
  c_events = country_dict[cname]['year'].value_counts().sort_index()
  c_success = c_yearly_sum[cname]['success'].sort_index()
  fig = line_plots([c_events, c_success], ['Number of terrorist events', 'Number of successful events'])
  fig.update_layout(title = f'Yearly terrorist events and successful terrorist events in {cname}', yaxis_title = 'Number of events', xaxis_title = 'Year',
                    autosize = False, width = 1300, height = 600)
  fig.show()

  yr_kw_list = [c_yearly_sum[cname][var] for var in ['killed', 'wounded']]
  fig = line_plots(yr_kw_list, ['Number of people killed', 'Number of people wounded'])
  fig.update_layout(title = f'Yearly casualties in {cname} due to terrorism', yaxis_title = 'Number of people', xaxis_title = 'Year',
                    autosize = False, width = 1300, height = 600)
  fig.show()

  c_groups = list(c_group_dict[cname].head(5).index)
  c_group_year = {group: country_dict[cname][country_dict[cname]['group'] == group]['year'].value_counts().sort_index() for group in c_groups}
  multi_cat_lineplot(c_groups, c_group_year, f'groups (top 5) in {cname}')

  for df, text in zip([c_attack_dict[cname], c_weapon_dict[cname], c_target_dict[cname], c_group_dict[cname][:10]], ['Attack Type', 'Weapon Type', 'Target Type', 'of top 10 Terrorist Groups']):
    fig = go.Figure(data = go.Pie(values = df.values, labels = df.index))
    fig.update_layout(title = f'Proportions of each {text} on total terrorist events in {cname}', autosize = False, width = 1300, height = 550)
    fig.show()

  suicide_plot = country_dict[cname]['suicide'].map({0: 'Non-suicide attack', 1: 'Suicide attack'}).value_counts()
  fig = go.Figure(data = go.Pie(values = suicide_plot.values, labels = suicide_plot.index))
  fig.update_layout(title = f'Percentage of suicide attacks in {cname}', autosize = False, width = 1300, height = 550)
  fig.show()

In [ ]:
generate_country_plots('Iraq')

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 8

In [ ]:
# Chart - 8 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 9

In [ ]:
# Chart - 9 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 10

In [ ]:
# Chart - 10 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 11

In [ ]:
# Chart - 11 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 12

In [ ]:
# Chart - 12 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 13

In [ ]:
# Chart - 13 visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 14 - Correlation Heatmap

In [ ]:
# Correlation Heatmap visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

#### Chart - 15 - Pair Plot

In [ ]:
# Pair Plot visualization code

##### 1. Why did you pick the specific chart?

Answer Here.

##### 2. What is/are the insight(s) found from the chart?

Answer Here

## **5. Solution to Business Objective**

#### What do you suggest the client to achieve Business Objective ?
Explain Briefly.

Answer Here.

# **Conclusion**

Write the conclusion here.

### ***Hurrah! You have successfully completed your EDA Capstone Project !!!***